File: WEEK4/bj4_2252.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const adjList = Array.from({ length: n + 1 }, () => []);
const inDegree = new Array(n + 1).fill(0);

for (let i = 1; i <= m; i++) {
    const [a, b] = input[i].split(' ').map(Number);
    adjList[a].push(b);
    inDegree[b]++;
}

const queue = [];
let head = 0;
for (let i = 1; i <= n; i++) {
    if (inDegree[i] === 0) {
        queue.push(i);
    }
}

const result = [];
while (head < queue.length) {
    const current = queue[head++];
    result.push(current);

    for (const neighbor of adjList[current]) {
        inDegree[neighbor]--;
        if (inDegree[neighbor] === 0) {
            queue.push(neighbor);
        }
    }
}

console.log(result.join(' '));

File: WEEK4/bj4_2252.py
import sys
from collections import deque

input = sys.stdin.readline
n, m = map(int, input().split())

adjList = [[] for _ in range(n + 1)]
inDegree = [0] * (n + 1)

for _ in range(m):
    a, b = map(int, input().split())
    adjList[a].append(b)
    inDegree[b] += 1

queue = deque()
for i in range(1, n + 1):
    if inDegree[i] == 0:
        queue.append(i)

result = []
while queue:
    current = queue.popleft()
    result.append(current)

    for neighbor in adjList[current]:
        inDegree[neighbor] -= 1
        if inDegree[neighbor] == 0:
            queue.append(neighbor)

print(*result)
